pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -71,6 +75,41 @@ type Registry struct {
 	Prefix string `toml:"prefix"`
 }
 
+// PullSource consists of an Endpoint and a Reference. Note that the reference is
+// rewritten according to the registries prefix and the Endpoint's location.
+type PullSource struct {
+	Endpoint  Endpoint
+	Reference reference.Named
+}
+
+// PullSourcesFromReference returns a slice of PullSource's based on the passed
+// reference.
+func (r *Registry) PullSourcesFromReference(ref reference.Named) ([]PullSource, error) {
+	var endpoints []Endpoint
+
+	if r.MirrorByDigestOnly {
+		// Only use mirrors when the reference is a digest one.
+		if _, isDigested := ref.(reference.Canonical); isDigested {
+			endpoints = append(r.Mirrors, r.Endpoint)
+		} else {
+			endpoints = []Endpoint{r.Endpoint}
+		}
+	} else {
+		endpoints = append(r.Mirrors, r.Endpoint)
```

Comment:
LGTM overall, one last non-blocking nit: Couldn’t this bit of logic also not be duplicated? E.g.
```go
if _, isDigested := ref.(reference.Canonical); isDigested || !r.MirrorByDigestOnly
```
is a bit ugly but IMHO still better than the duplication.

Feel free to merge this as it is if you prefer the current three-branch version.

Reply:
For whatever reason the comment did not show up in my emails. I thought about that but was hesitant as the cast is only necessary when MirrorByDigestOnly == true. I'll merge, so we can unblock the other pending PRs.